### DISCUSSION TITLE
Get image that is not returned in getImageList

### DISCUSSION THREAD

**Discussion by timurlenk07** [pymupdf/PyMuPDF#1202] [Looking for help]
I have a PDF document of an electronically generated invoice (which I am unfortunately unable to share), that I want to get OCR-ed by extracting all texts and running OCR on images. Unfortunately, the PDF has images that are not shown in`getImageList`, thus I cannot run OCR on it. Does someone have any idea, in what way could you embed an image without showing up either in PyMuPDF or Adobe Reader (in both cases, rendering the page results in the company logo shown but unable to find it/select it in the GUI) and how could I yet find it and extract its contents?

I know this is hard to answer without an actual sample, I will try to find something that I can share here, but in the meantime I would gladly accept any ideas about what the issue could be.

**Comment by timurlenk07** [comment]
Well I might have found part of the solution after a day of struggle. What I thought were images were actually *drawings*, so [this](https://pymupdf.readthedocs.io/en/latest/faq.html#extracting-drawings) recipe can extract them as a new PDF, while I have seen [other questions](https://github.com/pymupdf/PyMuPDF/discussions/1175#discussion-3486799) that try to extract it as SVG images, but that is still not answered.

Since I just want to "convert" the PDF to a textual form with bounding box information retained, is there any simple (simpler) and foulproof way to extract all the drawings into an image file?

I have also seen strange artifacts with the recipe I linked above, drawn elements that were not visible on the original PDF. What can be the cause of it and how can I get rid of it? I have seen that some elements are only visible when printing, but in this case a printed version looks exactly the same as the original, when viewed in Adobe reader.

**Comment by JorjMcKie** [comment]
>  thus I cannot run OCR on it

... this is an unjustified conclusion and may be true only with some OCR tools.
At least with [ocrmypdf](https://pypi.org/project/ocrmypdf/) you would nevertheless be able to OCR it.

**Reply by timurlenk07** [reply]
I meant that `getImagesList()` returns an empty list, because (as I have found out since) those images are actually *drawings*. Without an image, I don't have anything to run my own OCR engine on, that's what I've wanted to say.

BTW, I would be interested whether ocrmypdf can actually "convert" those drawings to a text layer, will check it out. Only problem though is then I would be forced to install tesseract.

**Reply by JorjMcKie** [reply]
ocrmypdf installs tesseract under the hood automatically, so if you do not have principal objections to tesseract, that should be not issue.
ocrmypdf has a number of options to force it OCR-ing a page with or without images, and with or without already present OCR-ed text.

**Reply by JorjMcKie** [reply]
After ocrmypdf, you can extract that text even including text position info using PyMuPDF. If you import ocrmypdf in your Python script, you can even integrate tesseract results with pymupdf text extractions in a fairly seemless way.

**Reply by timurlenk07** [reply]
I am trying it now. I would like to use my own OCR engine, that is why I don't really prefer installing ocrmypdf and using tesseract. Also, I want to do this "OCR" operation in-memory, since I want to process all the textual information visible on the rendered PDF.

**Reply by JorjMcKie** [reply]
> "OCR" operation in-memory

ocrmypdf does support that too.

Another option would always be to pixmap your difficult page (using a high resolution) and then feed it to your own OCR engine. Can all be done in memory, too.

**Reply by timurlenk07** [reply]
Another thing is that at the moment I also can speed up OCR using batch processing, which I am not sure is possible with ocrmypdf. So I would really prefer a simple solution to extract drawings as they appear in the rendered PDF.

**Reply by timurlenk07** [reply]
Well, I might delve a bit deeper into ocrmypdf then, before adding more stupid comments 😄 thanks for your help

**Reply by JorjMcKie** [reply]
> So I would really prefer a simple solution to extract drawings as they appear in the rendered PDF.

There are severe problems with this request: It would mean to detect, which object on the page covers which others (partly), taking also into account, the the "covering" object may be semi-transparent - - - endless problems which you could only resolve by re-inventing the better part of a PDF viewer.
At least PyMuPDF only extract text, image, drawings **_without_** caring about which covers what. You **_can_** of course look at the accompanying rectangles, e.g. within the list of drawing paths. Because their sequence is the same as the respective drawing command on the page, you can deduct some information from that.
The real difficulties arise, if a (non-transparent) drawing occurs after, say, some text which it covers: **_this_** cannot be detected right now.

**Reply by JorjMcKie** [reply]
> "... the "covering" object may be semi-transparent ..."

Forgot to add: **_.".. and have different colors"_** (!!!)
